libzecale/circuits/pairing/group_variable_gadgets.tcc

```diff
@@ -0,0 +1,439 @@
+        _dbl_gadgets.emplace_back(new dbl_gadget(
+            pb, *last_value, result, FMT(annotation_prefix, " double[0]")));
+    }
+}
+
+template<
+    typename groupT,
+    typename groupVariableT,
+    typename add_gadget,
+    typename dbl_gadget,
+    typename scalarT>
+void point_mul_by_const_scalar_gadget<
+    groupT,
+    groupVariableT,
+    add_gadget,
+    dbl_gadget,
+    scalarT>::generate_r1cs_constraints()
+{
+    const size_t last_bit = _scalar.num_bits() - 1;
+    size_t dbl_idx = 0;
+    size_t add_idx = 0;
+    for (ssize_t i = last_bit - 1; i >= 0; --i) {
+        // Double gadget constraints
+        _dbl_gadgets[dbl_idx++]->generate_r1cs_constraints();
+
+        // Add gadget constraints
+        if (_scalar.test_bit(i)) {
+            _add_gadgets[add_idx++]->generate_r1cs_constraints();
+        }
+    }
+}
+
+template<
+    typename groupT,
+    typename groupVariableT,
+    typename add_gadget,
+    typename dbl_gadget,
+    typename scalarT>
+void point_mul_by_const_scalar_gadget<
+    groupT,
+    groupVariableT,
+    add_gadget,
+    dbl_gadget,
+    scalarT>::generate_r1cs_witness()
+{
+    const size_t last_bit = _scalar.num_bits() - 1;
+    size_t dbl_idx = 0;
+    size_t add_idx = 0;
+    for (ssize_t i = last_bit - 1; i >= 0; --i) {
+        // Double gadget constraints
+        _dbl_gadgets[dbl_idx++]->generate_r1cs_witness();
+
+        // Add gadget constraints
+        if (_scalar.test_bit(i)) {
+            _add_gadgets[add_idx++]->generate_r1cs_witness();
+        }
+    }
+}
+
+template<
+    typename groupT,
+    typename groupVariableT,
+    typename add_gadget,
+    typename dbl_gadget,
+    typename scalarT>
+const groupVariableT &point_mul_by_const_scalar_gadget<
+    groupT,
+    groupVariableT,
+    add_gadget,
+    dbl_gadget,
+    scalarT>::result() const
+{
+    return _result;
+}
+
+// G2_add_gadget
+
+template<typename wppT>
+G2_add_gadget<wppT>::G2_add_gadget(
+    libsnark::protoboard<libff::Fr<wppT>> &pb,
+    const libsnark::G2_variable<wppT> &A,
+    const libsnark::G2_variable<wppT> &B,
+    const libsnark::G2_variable<wppT> &C,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<libff::Fr<wppT>>(pb, annotation_prefix)
+    , _A(A)
+    , _B(B)
+    , _C(C)
+    , _lambda(pb, FMT(annotation_prefix, " lambda"))
+    , _nu(pb, FMT(annotation_prefix, " nu"))
+    // lambda * (Bx - Ax) = By - Ay
+    , _lambda_constraint(
+          pb,
+          _lambda,
+          *_B.X - *_A.X,
+          *_B.Y - *_A.Y,
+          FMT(annotation_prefix, " _lambda_constraint"))
+    // lambda * Ax = Ay - nu
+    , _nu_constraint(
+          pb,
+          _lambda,
+          *_A.X,
+          *_A.Y - _nu,
+          FMT(annotation_prefix, " _nu_constraint"))
+    // lambda^2 = Cx + Ax + Bx
+    , _Cx_constraint(
+          pb,
+          _lambda,
+          _lambda,
+          *_C.X + *_A.X + *_B.X,
+          FMT(annotation_prefix, " _Cx_constraint"))
+    // Cx * lambda = -Cy - nu
```

Comment:
Usually, the "textbook" formulas for (affine) point addition on short Weierstrass curves are given as:
```
Let P=(Px, Py), Q=(Qx, Qy), then R=(Rx, Ry) = P+Q is obtained by:

\lambda = (y2−y1)/(x2−x1)
Rx = \lambda^2 - Px - Qx
Ry = \lambda * (Px - Rx) - Py
```

For simplicity and to make us a favor when we dig back into this code in 6 months, I'd either add a reference to the formulas you followed, or simply drop a few more lines in comments to explicit that the formulas you use in the constraint system are equivalent to those we usually find in textbooks (i.e. without the "nu")

Following your notations (C = A+B) such comment would be, for e.g.
```
Let A=(Ax, Ay), B=(Bx, By), then C=(Cx, Cy) = A+B is obtained by:

\lambda = (By - Ay)/(Bx - Ax)
Cx = \lambda^2 - Ax - Bx
Cy = \lambda * (Ax - Cx) - Ay
    = - (-\lambda * Ax + \lambda * Cx + Ay)
    = - (\lambda * Cx + Ay - \lambda * Ax)
    = - (\lambda * Cx + \nu), where \nu = Ay - \lambda * Ax
```

Reply:
Removed nu.  This also allows some constraints to be removed.

Reply:
Perfect